File: leetcode/106-从中序与后序遍历序列构造二叉树/106-从中序与后序遍历序列构造二叉树.go
package main

/**
* Definition for a binary tree node.

 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 106-从中序与后序遍历序列构造二叉树
func buildTree(inorder []int, postorder []int) *TreeNode {
	if len(postorder) == 0 || len(inorder) == 0 {
		return nil
	}
	root := &TreeNode{Val: postorder[len(postorder)-1]}
	index := 0
	for i, val := range inorder {
		if val == postorder[len(postorder)-1] {
			index = i
			break
		}
	}
	root.Left = buildTree(inorder[:index+1], postorder[:index+1])
	root.Right = buildTree(inorder[index+1:], postorder[index:len(postorder)-1])
	return root
}

File: leetcode/106-从中序与后序遍历序列构造二叉树/106-从中序与后序遍历序列构造二叉树.py
# Definition for a binary tree node.
from turtle import pos


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 106-从中序与后序遍历序列构造二叉树
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        if not inorder or not postorder:
            return None
        root = TreeNode(postorder[-1])
        index = inorder.index(root.val)

        root.left = self.buildTree(inorder[:index],postorder[:index])
        root.right = self.buildTree(inorder[index+1:],postorder[index:-1])
        return root


